apps/transport/lib/transport_web/plugs/block_user_agent.ex

```diff
@@ -0,0 +1,56 @@
+defmodule TransportWeb.Plugs.BlockUserAgent do
+  @moduledoc """
+  A Plug to:
+  - log the User-Agent making an HTTP request
+  - block HTTP requests when certain keywords are found in the User-Agent HTTP header
+
+  Options:
+  - log_user_agent: boolean or string
+  - block_user_agent_keywords: list of keywords or string that will be split on `|`
+  """
+  require Logger
+  import Plug.Conn
+  @behaviour Plug
+
+  @impl true
+  def init(log_user_agent: log_user_agent, block_user_agent_keywords: block_user_agent_keywords)
+      when is_binary(log_user_agent) and is_binary(block_user_agent_keywords) do
+    keywords = if block_user_agent_keywords == "", do: [], else: block_user_agent_keywords |> String.split("|")
+
+    init(
+      log_user_agent: log_user_agent == "true",
+      block_user_agent_keywords: keywords
+    )
+  end
+
+  def init(options) do
+    Keyword.validate!(options, log_user_agent: false, block_user_agent_keywords: [])
+  end
+
+  @impl true
+  def call(%Plug.Conn{request_path: request_path} = conn, options) do
+    if Keyword.fetch!(options, :log_user_agent) do
+      [user_agent] = get_req_header(conn, "user-agent")
+      Logger.info("Handling request #{request_path} by user-agent: #{user_agent}")
```

Comment:
Ceci est un commentaire non bloquant :)

J’ai testé, ça marche niquel :
```
[info] GET /
[info] Handling request / by user-agent: Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/119.0
[debug] Processing with TransportWeb.PageController.index/2
```

Je me demandais cela dit si on ne pouvait pas en profiter pour faire un plug qui fasse ça sur la même ligne que sur la requête entrante, et éventuellement avec l’IP, donc qui poste un truc du genre 
`[info] ip=1.2.3.4 user-agent=Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/119.0 action=GET route=/` à la place de juste `[info] GET /`.

Je suis tombé sur [ce truc là](https://stackoverflow.com/questions/39283529/logging-ip-addresses-for-all-visitors) qui choppe l’IP (à adapter pour le user-agent) et le passe sur la ligne suivante au plug Plug.Logger, qui est paramétré avec un format spécial, voir https://blog.appsignal.com/2023/07/18/observe-your-phoenix-app-with-structured-logging.html pour une explication plus longue de  `config :logger, :console, format:`.

Ce que je propose : merger ça, et donc avoir les logs des requêtes entrantes sur deux lignes, mais peut-être ensuite envisager de modifier le format des requêtes entrantes sur la base du blogpost de AppSignal ?

Reply:
Merci pour ta bonne suggestion !

J'ai fait ceci dans https://github.com/etalab/transport-site/pull/3572/commits/4a62950292417fad8a147db55b6592bd6816aaa4

Exemple en local :
> 13:57:20.546 request_id=F5Lj8oo7aIoicj4AAG_j remote_ip=127.0.0.1 http_method=GET http_path=/mission http_user_agent=Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/118.0.0.0 Safari/537.36 [info] Sent 404 in 222ms

J'ai demandé au support d'AppSignal pour que ce soit bien reconnu, on a bien `request_id` actuellement mais ce n'est pas détecté comme _log attribute_ avec notre drain Clever Cloud. Je demande si on doit adapter des choses côté AppSignal ou configuration.